Sources/Aperture/Aperture.swift

```diff
@@ -1,228 +1,846 @@
 
-	private let destination: URL
-	private let session: AVCaptureSession
-	private let output: AVCaptureMovieFileOutput
-	private var activity: NSObjectProtocol?
+		public var onStart: (() -> Void)?
+		public var onFinish: (() -> Void)?
+		public var onPause: (() -> Void)?
+		public var onResume: (() -> Void)?
 
-	public var onStart: (() -> Void)?
+		public var onError: ((Error) -> Void)? {
+			didSet {
+				recordingSession?.onError = onError
+			}
+		}
 
-	/**
-	Called when the recording finishes.
+		public var isPaused: Bool {
+			recordingSession?.isPaused ?? false
+		}
 
-	The success case may contain an error indicating a warning. It's recommended to present this to the user while informing them that the recording successfully finished.
-	*/
-	public var onFinish: ((Result<Swift.Error?, Swift.Error>) -> Void)?
+		public func startRecording(
+			target: Target,
+			options: RecordingOptions
+		) async throws {
+			guard recordingSession == nil else {
+				throw Error.recorderAlreadyStarted
+			}
 
-	public var onPause: (() -> Void)?
-	public var onResume: (() -> Void)?
-	public var isRecording: Bool { output.isRecording }
-	public var isPaused: Bool { output.isRecordingPaused }
+			let recordingSession = RecordingSession()
+			recordingSession.onError = onError
+			try await recordingSession.startRecording(target: target, options: options)
+			self.recordingSession = recordingSession
+			onStart?()
+		}
 
-	private init(
-		destination: URL,
-		input: AVCaptureInput,
-		output: AVCaptureMovieFileOutput,
-		audioDevice: AVCaptureDevice? = nil,
-		videoCodec: AVVideoCodecType? = nil
-	) throws {
-		self.destination = destination
-		self.session = AVCaptureSession()
-		self.output = output
+		public func stopRecording() async throws {
+			guard let recordingSession else {
+				throw Error.recorderNotStarted
+			}
 
-		// Needed because otherwise there is no audio on videos longer than 10 seconds.
-		// https://stackoverflow.com/a/26769529/64949
-		output.movieFragmentInterval = .invalid
+			try await recordingSession.stopRecording()
+			self.recordingSession = nil
+			onFinish?()
+		}
 
-		if let audioDevice = audioDevice {
-			if !audioDevice.hasMediaType(.audio) {
-				throw Error.invalidAudioDevice
+		public func pause() throws {
+			guard let recordingSession else {
+				throw Error.recorderNotStarted
 			}
 
-			let audioInput = try AVCaptureDeviceInput(device: audioDevice)
+			recordingSession.pause()
+			onPause?()
+		}
 
-			if session.canAddInput(audioInput) {
-				session.addInput(audioInput)
-			} else {
-				throw Error.couldNotAddMic
+		public func resume() async throws {
+			guard let recordingSession else {
+				throw Error.recorderNotStarted
 			}
+
+			await recordingSession.resume()
+			onResume?()
 		}
+	}
+}
 
-		if session.canAddInput(input) {
-			session.addInput(input)
-		} else {
-			throw Error.couldNotAddScreen
+extension Aperture {
+	internal final class RecordingSession: NSObject {
```

Comment:
```suggestion
	final class RecordingSession: NSObject {
```

`internal` is the default.